test/src/folding/foldable_block_test.dart

```diff
@@ -0,0 +1,148 @@
+      const blocks = [
+        FoldableBlock(startLine: 0, endLine: 1, type: FBT.braces),
+      ];
+
+      final actual = blocks.deepCopy()..joinIntersecting();
+
+      expect(actual, blocks);
+    });
+
+    test('Join intersecting blocks in list with two intersecting blocks', () {
+      const blocks = [
+        FoldableBlock(startLine: 0, endLine: 1, type: FBT.braces),
+        FoldableBlock(startLine: 1, endLine: 2, type: FBT.braces),
+      ];
+      const matcher = [
+        FoldableBlock(startLine: 0, endLine: 2, type: FBT.union),
+      ];
+
+      final actual = blocks.deepCopy()..joinIntersecting();
+
+      expect(actual, matcher);
+    });
+
+    test('Join intersecting blocks in list with two not intersecting blocks',
+        () {
+      const blocks = [
+        FoldableBlock(startLine: 0, endLine: 1, type: FBT.braces),
+        FoldableBlock(startLine: 2, endLine: 3, type: FBT.braces),
+      ];
+
+      final actual = blocks.deepCopy()..joinIntersecting();
+
+      expect(actual, blocks);
+    });
+
+    test('Multiple not intersected blocks will not join', () {
+      const blocks = [
+        FB(startLine: 0, endLine: 3, type: FBT.braces),
+        FB(startLine: 4, endLine: 5, type: FBT.imports),
+        FB(startLine: 8, endLine: 10, type: FBT.parentheses),
+      ];
+
+      final actual = blocks.deepCopy()..joinIntersecting();
+
+      expect(actual, blocks);
+    });
+
+    test('Multiple intersected blocks will join', () {
+      const blocks = [
+        FB(startLine: 0, endLine: 3, type: FBT.braces),
+        FB(startLine: 2, endLine: 5, type: FBT.imports),
+        FB(startLine: 4, endLine: 10, type: FBT.parentheses),
+      ];
+      const matcher = [
+        FB(startLine: 0, endLine: 10, type: FBT.union),
+      ];
+
+      final actual = blocks.deepCopy()..joinIntersecting();
+
+      expect(actual, matcher);
+    });
+
+    test('Mixed blocks will join correctly', () {
+      const blocks = [
+        FB(startLine: 0, endLine: 3, type: FBT.braces),
+        FB(startLine: 2, endLine: 5, type: FBT.imports),
+        FB(startLine: 6, endLine: 10, type: FBT.parentheses),
+      ];
+      const matcher = [
+        FB(startLine: 0, endLine: 5, type: FBT.union),
+        FB(startLine: 6, endLine: 10, type: FBT.parentheses),
+      ];
+
+      final actual = blocks.deepCopy()..joinIntersecting();
+
+      expect(actual, matcher);
+    });
+
+    test('Nested intersected blocks will join', () {
+      const blocks = [
+        FB(startLine: 0, endLine: 5, type: FBT.braces),
+        FB(startLine: 1, endLine: 3, type: FBT.imports),
+        FB(startLine: 3, endLine: 5, type: FBT.parentheses),
+      ];
+      const matcher = [
+        FB(startLine: 0, endLine: 5, type: FBT.braces),
+        FB(startLine: 1, endLine: 5, type: FBT.union),
+      ];
+
+      final actual = blocks.deepCopy()..joinIntersecting();
+
+      expect(actual, matcher);
+    });
+
+    test('Unsorted intersected blocks will not join', () {
```

Comment:
I don't think we should test this case at all. We may keep the behavior on unsorted lists undefined. Then it will be easier to streamline the algorithm on sorted input.

Please add this expectation to the doc comment of the method.